source/common/http/codes.cc

```diff
@@ -18,111 +18,132 @@
-  const uint64_t response_code = info.response_status_code_;
-  chargeBasicResponseStat(info.cluster_scope_, info.prefix_, static_cast<Code>(response_code));
+  Code code = static_cast<Code>(info.response_status_code_);
+
+  ASSERT(&info.cluster_scope_.symbolTable() == &symbol_table_);
+  chargeBasicResponseStat(info.cluster_scope_, info.prefix_, code);
 
-  std::string group_string =
-      CodeUtility::groupStringForResponseCode(static_cast<Code>(response_code));
+  Stats::StatName rq_group = upstreamRqGroup(code);
+  Stats::StatName rq_code = upstream_rq_.statName(code);
+
+  auto write_category = [this, rq_group, rq_code, &info](Stats::StatName category) {
+    incCounter(info.cluster_scope_, {info.prefix_, category, upstream_rq_completed_});
+    incCounter(info.cluster_scope_, {info.prefix_, category, rq_group});
+    incCounter(info.cluster_scope_, {info.prefix_, category, rq_code});
+  };
 
   // If the response is from a canary, also create canary stats.
   if (info.upstream_canary_) {
-    info.cluster_scope_.counter(absl::StrCat(info.prefix_, canary_upstream_rq_completed_)).inc();
-    info.cluster_scope_.counter(absl::StrCat(info.prefix_, canary_upstream_rq_, group_string))
-        .inc();
-    info.cluster_scope_.counter(absl::StrCat(info.prefix_, canary_upstream_rq_, response_code))
-        .inc();
+    write_category(canary_);
   }
 
   // Split stats into external vs. internal.
   if (info.internal_request_) {
-    info.cluster_scope_.counter(absl::StrCat(info.prefix_, internal_upstream_rq_completed_)).inc();
-    info.cluster_scope_.counter(absl::StrCat(info.prefix_, internal_upstream_rq_, group_string))
-        .inc();
-    info.cluster_scope_.counter(absl::StrCat(info.prefix_, internal_upstream_rq_, response_code))
-        .inc();
+    write_category(internal_);
   } else {
-    info.cluster_scope_.counter(absl::StrCat(info.prefix_, external_upstream_rq_completed_)).inc();
-    info.cluster_scope_.counter(absl::StrCat(info.prefix_, external_upstream_rq_, group_string))
-        .inc();
-    info.cluster_scope_.counter(absl::StrCat(info.prefix_, external_upstream_rq_, response_code))
-        .inc();
+    write_category(external_);
   }
 
   // Handle request virtual cluster.
   if (!info.request_vcluster_name_.empty()) {
-    info.global_scope_
-        .counter(join({vhost_, info.request_vhost_name_, vcluster_, info.request_vcluster_name_,
-                       upstream_rq_completed_}))
-        .inc();
-    info.global_scope_
-        .counter(join({vhost_, info.request_vhost_name_, vcluster_, info.request_vcluster_name_,
-                       absl::StrCat(upstream_rq_, group_string)}))
-        .inc();
-    info.global_scope_
-        .counter(join({vhost_, info.request_vhost_name_, vcluster_, info.request_vcluster_name_,
-                       absl::StrCat(upstream_rq_, response_code)}))
-        .inc();
+    Stats::StatName vhost_name = info.request_vhost_name_;
```

Comment:
nit: are the local variables needed? Same below?

Reply:
Not needed anymore; inlined. They were needed in a prior state of the PR when the info.request_zone_name_ was a string and I had to make a StatName for it here in the hot path.